### DISCUSSION TITLE
Guidance on WGSL Pointers as Function Parameters

### DISCUSSION THREAD

**Discussion by mrshannon** [gpuweb/gpuweb#2197] [Q&A]
In GLSL (which is what most WebGPU users are likely most familiar with) there are `in`, `out`, and `inout` parameters. There are also return values which are basically the same as using `out` on a function parameter. WGSL instead has pass by value (or is this really a reference), pass by pointer, and return values. Since GLSL does not have pointers, when they should be used is not very clear, especially with respect to function parameters.

Therefore, what should the signature be in WGSL for a GLSL function with the following signature:
```
bool myFunction(in vec3 a, inout vec3 b, out vec3 c)
```

So there are three cases:

- `in` - my guess is pass by value, through you could pass by pointer
- `inout` - pass by pointer seems to be the only way to replicate this in WGSL
- `out` - specifically regarding multiple outputs, could pack into a struct and return, or could again use pass by pointer

I am guessing `inout` and `out` should always translate to pointers. Is this true?

As a C programmer I also want to say that `in` should translate to a pointer for anything besides `f32`, `u32`, and `i32` but I am guessing this is likely the wrong mental model in WGSL.

_It also seems strange that a pointer parameter must specify `private` or `function` scope as that limits the generality of functions which take parameters via pointers. But I am guessing there is some sort of underlying restriction. Would be nice for this to be deduced at compile time though._

**Comment by kvark** [comment]
Converting from GLSL to WGSL (e.g. using [naga](https://github.com/gfx-rs/naga)) could be the starting point.
E.g. this source
```glsl
bool myFunction(in vec3 a, inout vec3 b, out vec3 c) {
	c = a;
	b *= 2.0;
	return true;
}
...
```
Is translated to
```rust
fn myFunction(a: vec3<f32>, b: ptr<function, vec3<f32>>, c: ptr<function, vec3<f32>>) -> bool {
    var a_1: vec3<f32>;
    a_1 = a;
    let _e4 = a_1;
    (*c) = _e4;
    let _e5 = (*b);
    (*b) = (_e5 * 2.0);
    return true;
}
```

This is hardly a guidance, just posting for the reference.
